### ISSUE TITLE
Building locally

### ISSUE THREAD

**Issue by skosch** [boldpoker/edeliver#134]
Hi – this may be a stupid question borne out of my lack of experience with CI tools, but I'd like to use `edeliver` without a build server and staging server. Is this possible?

Due to, erm, institutional constraints on the customer end, I only have access to *one* server and *one* DB, all tightly integrated with their single-sign-on infrastructure; plus I'm the only one on the job. I'd like to use edeliver only to build my package locally (no git pushing etc.) and to then copy it to the production server. Is this possible? Or am I better off just doing it manually?

**Comment by dustinfarris** [comment]
You can build on any machine you want—your local, the production server, or a separate build server.  Just point edeliver to it in the config.  Keep in mind that you want the build environment and the production environment to be as similar as possible since the build will be compiling your code.

You can supply the same target for both staging and production.

**Comment by skosch** [comment]
When trying to build locally, I get
```
BUILDING RELEASE OF APP ON BUILD HOST

-----> Authorizing hosts
-----> Ensuring hosts are ready to accept git pushes
ssh: connect to host localhost port 22: Connection refused
```
I took that to mean that to build, code had to be git-pushed to a machine running a git server. Can I instruct edeliver to just grab code from a local directory?

**Comment by bharendt** [comment]
You don't need a git-server running on the (local) build host. edeliver uses ssh to push your (locally committed) changes to the build host. If you want to use your localhost as build host you must ensure that sshd is running and that you can login to your localhost, e.g. by creating a keypair (if you don't have any) like this: 

```bash
ssh-keygen -t rsa -b 4096
```

your edeliver config should then use localhost as build host:

```bash
# .deliver/config
BUILD_HOST="localhost"
BUILD_USER="$USER"
BUILD_AT="~/my-build-dir"
```

As @dustinfarris already mentioned, this only makes sense if your development environment uses the same os / architecture as your deploy environment because the erts (erlang runtime) is included into the release and native / system-dependend code might be included as well.

Another option would be to use a local docker container as build host, e.g. like this:

```
# builder.dockerfile
FROM msaraiva/elixir-dev:1.3.1
RUN apk add --update grep autoconf alpine-sdk ncurses ncurses-dev openssl openssl-dev bash git curl erlang-tools openssh \
    && rm -rf /var/cache/apk/*

RUN sed -i 's/^#PermitRootLogin prohibit-password/PermitRootLogin yes/' /etc/ssh/sshd_config
RUN sed -i 's/^#PasswordAuthentication yes/PasswordAuthentication no/' /etc/ssh/sshd_config

RUN ssh-keygen -t ed25519 -f /etc/ssh/ssh_host_ed25519_key -N ''
RUN ssh-keygen -t rsa     -f /etc/ssh/ssh_host_rsa_key     -N ''
RUN ssh-keygen -t dsa     -f /etc/ssh/ssh_host_dsa_key     -N ''
RUN ssh-keygen -t ecdsa   -f /etc/ssh/ssh_host_ecdsa_key   -N ''

ENV TERM vt100
VOLUME /my-build-dir
CMD /usr/sbin/sshd -D
```

... build it

```bash
docker build -t my-repository/builder -f builder.dockerfile
```

... and start it before building with edeliver 

```bash
docker run -d -t my-repository/builder --expose=22 --volume=~/.ssh/id_rsa.pub:/root/.ssh/authorized_keys --name=builder 
```

Your edeliver config should then look like this:

```bash
# .deliver/config
BUILD_HOST="localhost"
BUILD_USER="root"
BUILD_AT="~/my-build-dir"
```

**Comment by skosch** [comment]
Ah, so sshd is all I need. That explains the port 22 as well. Thanks!

**Comment by parmar7725274** [comment]
Failed to Build locally. ( sshd is running ) 

_error_ :

$ ./edeliver build release
fatal: Not a git repository (or any of the parent directories): .git

BUILDING RELEASE OF NEUTRON_EDELIVER APP ON BUILD HOST

-----> Authorizing hosts
fatal: Not a git repository (or any of the parent directories): .git
-----> Ensuring hosts are ready to accept git pushes
fatal: Not a git repository (or any of the parent directories): .git
utl@localhost's password: 
error: could not lock config file .git/config: No such file or directory

A remote command failed on:

  utl@localhost

Output of the command is shown above and the command executed
on that host is printed below for debugging purposes:

FAILED with exit status 255:

    set -e
    if [ ! -d /home/utl/edeliver_experiments ]
    then
      mkdir -p /home/utl/edeliver_experiments
      cd /home/utl/edeliver_experiments
      git init &> /dev/null
      git config receive.denyCurrentBranch ignore
    else
      cd /home/utl/edeliver_experiments
      git config receive.denyCurrentBranch ignore
    fi


Any solution ? I don't want to use git at all. Is there any option to tell edeliver to not to perform any git operation ?

**Comment by bobzhen** [comment]
@parmar7725274 Got any solution? I have same issue like yours....

**Comment by hopewise** [comment]
I have the same issue here, I tried to run the command on the remote server:

    git config receive.denyCurrentBranch ignore
I got this error:

    error: could not lock config file .git/config: No such file or directory

So, I think if we need to fix the issue on the remote server

**Comment by DavidOliver** [comment]
@parmar7725274, @bobzhen, @hopewise, when using localhost to build I also found that the Git repo failed to be cloned/pushed to the BUILT_AT directory. The readme says:

> It may be required to install and configure git on your build host. You may also have to clone the repository initially at the BUILD_AT path, although edeliver will try to take care of this for you.

Once I cloned my repo into the BUILD_AT directory, edeliver successfully proceeded past the Git stage. The terminal output included 'Pushing new commits with git to: username@localhost', so hopefully the BUILD_AT directory will be updated for subsequent builds.
